File: C/Searching-Methods/Binary_Search.c
#include<stdio.h>

int IBinarySearch(int A[], int n, int key);
int RBinarySearch(int A[], int low, int high, int key);

int main()
{
    printf("-------- BINARY SEARCH --------- \n\n");
    int n, arr[50], key, foundIndex;

    printf("Enter no. of values: ");
    scanf("%d", &n);

    printf("Enter array of values (with spaces and in sorted manner): ");
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &arr[i]);
    }

    printf("Enter the key for searching: ");
    scanf("%d", &key);

    // foundIndex = IBinarySearch(arr, n, key);
    int low=0, high=n-1;
    foundIndex = RBinarySearch(arr, low, high, key);

    if (foundIndex != -1)
    {
        printf("Element found at position %d\n", foundIndex+1);
    }
    else
        printf("Element not found\n");

    printf("\n");
    return 0;
}

int IBinarySearch(int A[], int n, int key)
{
    int low=0, high=n-1, mid=-1;

    while(low <= high) 
    {
        mid = (low + high)/2;

        if (key == A[mid]) return mid;
        else if (key < A[mid]) high = mid - 1;
        else low = mid + 1;
    }

    return -1;
}

int RBinarySearch(int A[], int low, int high, int key)
{
    int mid;

    if (low <= high)
    {
        mid = (low + high)/2;

        if (key == A[mid]) return mid;
        else if(key < A[mid]) return RBinarySearch(A, low, mid-1, key);
        else return RBinarySearch(A, mid+1, high, key);
    }

    return -1;
}

File: Javascript/Searching/BinarySearch.js
function IBinarySearch(arr, key) {
  let low = 0,
    high = arr.length - 1,
    mid = -1

  while (low <= high) {
    mid = Math.floor((low + high) / 2)

    if (key === arr[mid]) return mid
    else if (key < arr[mid]) high = mid - 1
    else low = mid + 1
  }

  return -1
}

/* Tail recursion used - can be coverted to iteration easily */
/* Since recursion uses stack so iteration is quite efficient here */
function RBinarySearch(arr, low, high, key) {
  if (low <= high) {
    let mid = Math.floor((low + high) / 2)

    if (key === arr[mid]) return mid
    else if (key < arr[mid]) return RBinarySearch(arr, low, mid - 1, key)
    else return RBinarySearch(arr, mid + 1, high, key)
  }

  return -1
}

const arr = [10, 20, 30, 40, 50, 60, 70, 80]

const key = 50

console.log("Element found at position: ", IBinarySearch(arr, key) + 1)

console.log(
  "Element found at position:",
  RBinarySearch(arr, 0, arr.length - 1, key) + 1
)
